system-tests/lib/cre/environment/environment.go

```diff
@@ -216,46 +223,81 @@ func SetupTestEnvironment(
 	}
 
 	for i, donMetadata := range topology.DonsMetadata {
-		config, configErr := keystoneporconfig.GenerateConfigs(
-			keystonetypes.GeneratePoRConfigsInput{
-				DonMetadata:                 donMetadata,
-				BlockchainOutput:            blockchainsOutput.BlockchainOutput,
-				DonID:                       donMetadata.ID,
-				Flags:                       donMetadata.Flags,
-				PeeringData:                 peeringData,
-				CapabilitiesRegistryAddress: keystoneContractsOutput.CapabilitiesRegistryAddress,
-				WorkflowRegistryAddress:     keystoneContractsOutput.WorkflowRegistryAddress,
-				ForwarderAddress:            keystoneContractsOutput.ForwarderAddress,
-				GatewayConnectorOutput:      topology.GatewayConnectorOutput,
-			},
-		)
-		if configErr != nil {
-			return nil, pkgerrors.Wrap(configErr, "failed to generate config")
+		configsFound := 0
+		secretsFound := 0
+		for _, nodeSpec := range input.CapabilitiesAwareNodeSets[i].NodeSpecs {
+			if nodeSpec.Node.TestConfigOverrides != "" {
+				configsFound++
+			}
+			if nodeSpec.Node.TestSecretsOverrides != "" {
+				secretsFound++
+			}
 		}
-
-		secretsInput := &keystonetypes.GenerateSecretsInput{
-			DonMetadata: donMetadata,
+		if configsFound != 0 && configsFound != len(input.CapabilitiesAwareNodeSets[i].NodeSpecs) {
+			return nil, fmt.Errorf("%d out of %d node specs have config overrides. Either provide overrides for all nodes or none at all", configsFound, len(input.CapabilitiesAwareNodeSets[i].NodeSpecs))
 		}
 
-		if evmKeys, ok := keys.EVMKeys[donMetadata.ID]; ok {
-			secretsInput.EVMKeys = evmKeys
+		if secretsFound != 0 && secretsFound != len(input.CapabilitiesAwareNodeSets[i].NodeSpecs) {
+			return nil, fmt.Errorf("%d out of %d node specs have secrets overrides. Either provide overrides for all nodes or none at all", secretsFound, len(input.CapabilitiesAwareNodeSets[i].NodeSpecs))
 		}
 
-		if p2pKeys, ok := keys.P2PKeys[donMetadata.ID]; ok {
-			secretsInput.P2PKeys = p2pKeys
+		// Allow providing only secrets, because we can decode them and use them to generate configs
+		// We can't allow providing only configs, because we can't replace secret-related values in the configs
+		// If both are provided, we assume that the user knows what they are doing and we don't need to validate anything
+		// And that configs match the secrets
+		if configsFound > 0 && secretsFound == 0 {
+			return nil, fmt.Errorf("nodese config overrides are provided for DON %d, but not secrets. You need to either provide both, only secrets or nothing at all", donMetadata.ID)
 		}
 
-		// EVM and P2P keys will be provided to nodes as secrets
-		secrets, secretsErr := keystonesecrets.GenerateSecrets(
-			secretsInput,
-		)
-		if secretsErr != nil {
-			return nil, pkgerrors.Wrap(secretsErr, "failed to generate secrets")
+		// generate configs only if they are not provided
+		if configsFound == 0 {
```

Comment:
For the sake of simplicity, is it possible to use static config or secrets and only have one case for the ephemeral environment? If it's not possible for config, what about just secrets? This way we can make it work the same for ephemeral and real environments by using different configs, one with test keys and another one with stage keys, etc?

Reply:
I would keep it for now as-is, because we do have some use cases for generating keys dynamically. For example sandbox environments running on private ephemeral chains or load tests, where we want to be able to easily change number of DONs/nodes without having to provide TOML config entry for each node (together with somehow generating the secrets for them).